src/transformStateWithClones.js

```diff
@@ -6,8 +6,34 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
 
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+
+    stateHistory.push(newState);
+    currentState = newState;
```

Comment:
You are creating a new state copy within the loop before the switch statement. To adhere to the DRY principle, you should create this copy only once after the switch statement, reflecting all the changes made by the current action.